File: examples/python-example/user_property_ref.py
"""
Example PropertyRef enum for User entity.
This shows how developers should define their own property references.
They only need to extend PropertyRef and call super() - all methods are inherited!
"""

from dynamic_filter_core.validation import PropertyRef, Operator


class UserPropertyRef(PropertyRef):
    """Example PropertyRef enum for User entity."""
    
    # User entity properties
    USER_NAME = PropertyRef("userName", "str", [Operator.LIKE, Operator.EQUALS, Operator.IN, Operator.NOT_IN])
    USER_EMAIL = PropertyRef("email", "str", [Operator.LIKE, Operator.EQUALS, Operator.IN, Operator.NOT_IN])
    USER_STATUS = PropertyRef("status", "str", [Operator.EQUALS, Operator.NOT_EQUALS, Operator.IN, Operator.NOT_IN])
    USER_AGE = PropertyRef("age", "int", [Operator.EQUALS, Operator.NOT_EQUALS, Operator.GREATER_THAN, Operator.GREATER_THAN_OR_EQUAL, Operator.LESS_THAN, Operator.LESS_THAN_OR_EQUAL, Operator.IN, Operator.NOT_IN, Operator.BETWEEN, Operator.NOT_BETWEEN])
    USER_CREATED_DATE = PropertyRef("createdDate", "datetime", [Operator.EQUALS, Operator.NOT_EQUALS, Operator.GREATER_THAN, Operator.GREATER_THAN_OR_EQUAL, Operator.LESS_THAN, Operator.LESS_THAN_OR_EQUAL, Operator.BETWEEN, Operator.NOT_BETWEEN])
    USER_IS_ACTIVE = PropertyRef("active", "bool", [Operator.EQUALS, Operator.NOT_EQUALS])
    USER_SALARY = PropertyRef("salary", "float", [Operator.EQUALS, Operator.NOT_EQUALS, Operator.GREATER_THAN, Operator.GREATER_THAN_OR_EQUAL, Operator.LESS_THAN, Operator.LESS_THAN_OR_EQUAL, Operator.BETWEEN, Operator.NOT_BETWEEN])

    @classmethod
    def find_by_entity_field(cls, entity_field: str) -> PropertyRef:
        """
        Finds a UserPropertyRef by its entity field name.
        
        Args:
            entity_field: The entity field name to search for
            
        Returns:
            The matching UserPropertyRef or None if not found
        """
        all_properties = [
            cls.USER_NAME,
            cls.USER_EMAIL,
            cls.USER_STATUS,
            cls.USER_AGE,
            cls.USER_CREATED_DATE,
            cls.USER_IS_ACTIVE,
            cls.USER_SALARY
        ]

        for prop in all_properties:
            if prop.entity_field == entity_field:
                return prop
        return None

    @classmethod
    def get_all_properties(cls) -> list[PropertyRef]:
        """
        Gets all UserPropertyRef values.
        
        Returns:
            A list of all UserPropertyRef values
        """
        return [
            cls.USER_NAME,
            cls.USER_EMAIL,
            cls.USER_STATUS,
            cls.USER_AGE,
            cls.USER_CREATED_DATE,
            cls.USER_IS_ACTIVE,
            cls.USER_SALARY
        ]

File: examples/csharp-example/UserPropertyRef.cs
using DynamicFilter.Core.Validation;

namespace DynamicFilter.Example
{
    /// <summary>
    /// Example PropertyRef enum for User entity.
    /// This shows how developers should define their own property references.
    /// They only need to extend PropertyRef and call super() - all methods are inherited!
    /// </summary>
    public class UserPropertyRef : PropertyRef
    {
        // User entity properties
        public static readonly UserPropertyRef UserName = new UserPropertyRef("userName", "string", new[] { Operator.Like, Operator.Equals, Operator.In, Operator.NotIn });
        public static readonly UserPropertyRef UserEmail = new UserPropertyRef("email", "string", new[] { Operator.Like, Operator.Equals, Operator.In, Operator.NotIn });
        public static readonly UserPropertyRef UserStatus = new UserPropertyRef("status", "string", new[] { Operator.Equals, Operator.NotEquals, Operator.In, Operator.NotIn });
        public static readonly UserPropertyRef UserAge = new UserPropertyRef("age", "int", new[] { Operator.Equals, Operator.NotEquals, Operator.GreaterThan, Operator.GreaterThanOrEqual, Operator.LessThan, Operator.LessThanOrEqual, Operator.In, Operator.NotIn, Operator.Between, Operator.NotBetween });
        public static readonly UserPropertyRef UserCreatedDate = new UserPropertyRef("createdDate", "DateTime", new[] { Operator.Equals, Operator.NotEquals, Operator.GreaterThan, Operator.GreaterThanOrEqual, Operator.LessThan, Operator.LessThanOrEqual, Operator.Between, Operator.NotBetween });
        public static readonly UserPropertyRef UserIsActive = new UserPropertyRef("active", "bool", new[] { Operator.Equals, Operator.NotEquals });
        public static readonly UserPropertyRef UserSalary = new UserPropertyRef("salary", "double", new[] { Operator.Equals, Operator.NotEquals, Operator.GreaterThan, Operator.GreaterThanOrEqual, Operator.LessThan, Operator.LessThanOrEqual, Operator.Between, Operator.NotBetween });

        private UserPropertyRef(string entityField, string type, Operator[] supportedOperators) 
            : base(entityField, type, supportedOperators)
        {
        }

        /// <summary>
        /// Finds a UserPropertyRef by its entity field name.
        /// </summary>
        /// <param name="entityField">The entity field name to search for</param>
        /// <returns>The matching UserPropertyRef or null if not found</returns>
        public static UserPropertyRef FindByEntityField(string entityField)
        {
            var allProperties = new[]
            {
                UserName,
                UserEmail,
                UserStatus,
                UserAge,
                UserCreatedDate,
                UserIsActive,
                UserSalary
            };

            return allProperties.FirstOrDefault(prop => prop.EntityField == entityField);
        }

        /// <summary>
        /// Gets all UserPropertyRef values.
        /// </summary>
        /// <returns>An array of all UserPropertyRef values</returns>
        public static UserPropertyRef[] GetAllProperties()
        {
            return new[]
            {
                UserName,
                UserEmail,
                UserStatus,
                UserAge,
                UserCreatedDate,
                UserIsActive,
                UserSalary
            };
        }
    }
}
